modules/core/arrow-core/src/main/kotlin/arrow/core/extensions/listk.kt

```diff
@@ -339,3 +340,33 @@ interface ListKUnalign : Unalign<ForListK>, ListKSemialign {
       }.bimap({ it.k() }, { it.k() })
     }
 }
+
+@extension
+interface ListKCrosswalk : Crosswalk<ForListK>, ListKFunctor, ListKFoldable {
+  override fun <F, A, B> crosswalk(
+    ALIGN: Align<F>,
+    a: Kind<ForListK, A>,
+    fa: (A) -> Kind<F, B>
+  ): Kind<F, Kind<ForListK, B>> =
+    a.fix().let { list ->
+      if (list.isEmpty()) {
+        ALIGN.run { empty<B>().map { ListK.empty<B>() } }
+      } else {
+        val head = list.first()
+        val tail = list.drop(1).k()
+
+        val ls: Kind<F, B> = fa(head)
+        val rs: Kind<F, Kind<ForListK, B>> = crosswalk(ALIGN, tail, fa)
+
+        ALIGN.run {
+          alignWith(ls, rs) { ior: Ior<B, Kind<ForListK, B>> ->
+            when (ior) {
+              is Ior.Left -> ListK.just(ior.value)
+              is Ior.Right -> ior.value
+              is Ior.Both -> ListK.just(ior.leftValue) + ior.rightValue.fix()
+            }
+          }
+        }
+      }
```

Comment:
This looks wrong. Not in the sense that the implementation might be wrong, but I feel like there has to be a better/cleaner way to write this. I'll take a look at this later with a bit more time :)

Reply:
looking forward for it!

Reply:
i would be really curious here to hear what you are looking for in a good FP implementation.as, coming from an OO background this would be really interesting for me to hear. so it would be great if you could share a bit of your thought process regarding the current implementation and the enhanced one.

Reply:
Main motivation is usually just to avoid _magic_ as much as possible and work on pure values.
That usually just involves avoiding language operators like `if`, loops, `return`, `throw` as much as possible and stick to `fold` functions and pattern matching over `when` whenever possible. `if` can have good uses, but the other constructs should be avoided at all cost. There is nothing inherently wrong from a functional setting with this code. It just feels overly verbose and is build from lots of small operations that feel like they should have some more general primitve that makes this code more concise.

Reply:
That is for low level code, higher level design of functional programs is a bit more and also very much opinion based. Usually sticking to pure functions, functions over data and no-magic is a good route forward.

Reply:
I kind of wanna write more about design in functional programming, but so much of what I want to do is only possible in haskell :/ I usually only use kotlin when I am bound to the jvm by previous work or by libraries that I have to use...

Reply:
> I kind of wanna write more about design in functional programming, but so much of what I want to do is only possible in haskell :/ I usually only use kotlin when I am bound to the jvm by previous work or by libraries that I have to use...

Which is most of the time at work for me, so I still use it a lot :)

Reply:
thanks for the insight! i think i will learn a bit more about haskell :-)